Services/Services_bases/gpt_service/gpt.py

```diff
@@ -87,6 +119,111 @@ async def get_chat_completion(
+
+    def store_signal_history(
+        self,
+        exchange: str,
+        symbol: str,
+        time_frame: commons_enums.TimeFrames,
+        version: str,
+        signals_by_candle_open_time,
+    ):
+        tf = time_frame.value
+        for candle_open_time, signal in signals_by_candle_open_time.items():
+            self.stored_signals.set_node_at_path(
+                signal,
+                str,
+                [exchange, symbol, tf, version, candle_open_time]
+            )
+
+    def has_signal_history(
+        self,
+        exchange: str,
+        symbol: str,
+        time_frame: commons_enums.TimeFrames,
+        min_timestamp: float,
+        max_timestamp: float,
+        version: str
+    ):
+        for ts in (min_timestamp, max_timestamp):
+            if self._get_signal_from_stored_signals(
+                exchange, symbol, time_frame.value, version, self._get_open_candle_timestamp(time_frame, ts)
+            ) == "":
+                return False
+        return True
+
+    async def _fetch_and_store_history(
+        self, authenticator, exchange_name, symbol, time_frame, version, min_timestamp: float, max_timestamp: float
+    ):
+        signals_by_candle_open_time = await authenticator.get_gpt_signals_history(
+            exchange_name, symbol, time_frame,
+            self._get_open_candle_timestamp(time_frame, min_timestamp),
+            self._get_open_candle_timestamp(time_frame, max_timestamp),
+            version
+        )
+        if not signals_by_candle_open_time:
+            self.logger.error(
+                f"No ChatGPT signal history for {symbol} on {time_frame.value} for {exchange_name} with {version}. "
+                f"Please check {self._supported_history_url()} to get the list of supported signals history."
+            )
+        self.store_signal_history(
+            exchange_name, symbol, time_frame, version, signals_by_candle_open_time
+        )
+
+    async def fetch_gpt_history(
+        self, exchange_name: str, symbols: list, time_frames: list,
+        version: str, start_timestamp: float, end_timestamp: float
+    ):
+        authenticator = authentication.Authenticator.instance()
+        coros = [
+            self._fetch_and_store_history(
+                authenticator, exchange_name, symbol, time_frame, version, start_timestamp, end_timestamp
+            )
+            for symbol in symbols
+            for time_frame in time_frames
+            if not self.has_signal_history(exchange_name, symbol, time_frame, start_timestamp, end_timestamp, version)
+        ]
+        if coros:
+            await asyncio.gather(*coros)
+
+    def _get_open_candle_timestamp(self, time_frame: commons_enums.TimeFrames, base_timestamp: float):
+        tf_seconds = commons_enums.TimeFramesMinutes[time_frame] * commons_constants.MINUTE_TO_SECONDS
+        return base_timestamp - (base_timestamp % tf_seconds)
```

Comment:
Don't we have already this code somewhere else?

Reply:
added in commons